### PR TITLE
Fix pdist and cdist launch failure past 2^24 outputs on ROCm

### LINKED ISSUE #168868: Test: TestTorchDeviceType.test_pdist_norm_large
## Context
- Source file: `test/test_torch.py`
- Test class: `TestTorchDeviceType`
- Test name: `test_pdist_norm_large`
- Defined at line: 4066
- Parent issue: #168861
- Code reference: https://github.com/pytorch/pytorch/blob/main/test/test_torch.py

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @dllehr-amd

### PR DESCRIPTION
**Pull request by geozhai**
Fixes https://github.com/pytorch/pytorch/issues/168868

`pdist_forward_kernel_impl` and `cdist_kernel_impl` launch one 256-thread block per output element. HSA expresses a grid dimension in work-items rather than workgroups, and `rocminfo` reports a per-dimension `Grid Max Size` of 4294967295, so `gridDim.x * blockDim.x` has to fit in uint32. With `blockDim.x` of 256 that caps a launch at 2^24 outputs. Bisected on gfx90a and gfx950 under HIP 7.14, pdist at n=5793 (16776528 blocks) succeeds and n=5794 (16782321 blocks) raises `hipErrorInvalidConfiguration`; cdist crosses the same line at exactly 16777216 outputs. `test_pdist_norm_large` uses n=50000, which needs 1249975000 blocks or roughly 74x the limit, which is why it was skipped on ROCm. CUDA expresses `maxGridSize[0]` in blocks, so the same launch is legal there.

The failure mode depends on the HIP runtime rather than the architecture. HIP 7.14 rejects the oversized dispatch outright, while HIP 7.2 on the same MI355X silently produces wrong results with outputs left unwritten. Both are the same oversized launch, so both are fixed by bounding it: each launch caps the grid at 32 blocks per CU through a shared `dist_grid` helper and grid-strides over the remaining outputs. `cdist` is fixed alongside `pdist` because it has the identical bug and fails at 2^24 outputs for the same reason.

The alternative to capping the grid is to keep one block per output and spread the launch across a 2-D grid, which would preserve the existing block-to-output mapping exactly. The cap plus grid-stride loop was chosen instead because it bounds the launch on CUDA as well, keeps a single index expression in the kernel, and measured within a few percent of the previous cost.

### Test Plan

On gfx950 (MI355X, HIP 7.14):

```
python -m pytest test/test_torch.py -k "pdist or cdist" -v
python -m pytest test/test_ops.py -k "cdist or pdist" -q
python -m pytest test/test_autograd.py test/test_nn.py -k "cdist or pdist or pairwise" -q
```

27 passed / 9 skipped, 81 passed / 15 skipped / 5 xfailed, and 31 passed / 2 xfailed respectively. `test_pdist_norm_large_cuda` passes in 4.7s once it is no longer skipped on ROCm. Note that it cannot be exercised through `PYTORCH_TEST_WITH_ROCM`, since `TEST_WITH_ROCM` is derived from `torch.version.hip` and `skipIfRocm` therefore always skips it.

Forward output was also compared against CPU for p in {0, 0.5, 1, 2, 3, inf} above the old boundary, for batched cdist, and for degenerate shapes, plus n=50000 end to end.

Performance, 10-iteration averages on MI355X, before versus after:

```
pdist n=4096 m=128     4.220 ms -> 4.283 ms
pdist n=4096 m=8       3.683 ms -> 3.928 ms
pdist n=2048 m=128     1.063 ms -> 1.082 ms
cdist 2048x4096 m=128  9.468 ms -> 9.384 ms
cdist 2048x4096 m=8    8.155 ms -> 8.769 ms
cdist 1024x4096 m=128  2.643 ms -> 2.774 ms
```

Original investigation and reproduction by @geozhai, who identified the oversized launch grid and the ROCm skip that was masking it. Portions of the analysis in this description were drafted with an AI assistant (Claude Code).

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Review by jeffdaily — CHANGES_REQUESTED** [review]
Nice investigation, but a few things before this lands:

1. **Keep the DeviceSqrt.cuh cleanup, but stop crediting it as a fix.** I diffed the GCN ISA of the old `::sqrt`/`::sqrtf` specializations vs the new `std::sqrt` template (HIP 7.2, gfx90a/942/950): instruction streams are byte-identical. `device_sqrt<double>` already lowered to full fp64 (`v_rsq_f64` + Newton-Raphson) in both -- it was never fp32. So the "device_sqrt ran in fp32 on ROCm / 2^24 limit" root cause is inaccurate. The cleanup is worth keeping (the in-namespace `#include <math.h>` was ugly), but it generates identical code; please reword the PR description and the test comment to present it as a no-op cleanup, not the fix. The actual forward fix is entirely the integer index-correction loop, papering over a ~1-ulp ROCm-vs-NVIDIA fp64-sqrt rounding difference at row boundaries.

2. **The backward kernel still has the bug.** `pdist_backward_kernel_cuda_impl` does the same `device_sqrt`-based index inversion (DistanceKernel.cu:173-174) with no correction, so it produces silently-wrong gradients in the same large-`n` regime. Apply the same `row_start` correction there and add a backward test past the boundaries.

3. **Shared CUDA code re-tuned for a ROCm-only issue.** The oversized-grid drop is gfx950-specific, but the grid-stride rewrite + hardcoded `max_blocks = 4096` + extra `__syncthreads()` now hit every NVIDIA GPU, validated only on MI355X. Keep the index correction shared (both backends need it), but gate the grid cap behind `#if defined(USE_ROCM)` or size it from device properties, and drop the magic `4096`.

4. **Drop the trailing `__syncthreads()` and its comment.** `cuda_utils::BlockReduce` already opens with `__syncthreads()` ("prevent races when BlockReduces are called in a row"), which already serializes `agg_smem` reuse across grid-stride iterations, so it's redundant.

**Comment by geozhai** [comment]
Thanks for the review  @jeffdaily , my response to each of your suggestions.

1. I accept. My former hypothesis of FP32 device_sqrt and the 2^24 limit was wrong. I also attempted to diff between the ISA of the original and modified code and can confirm your observations.

2. I cannot verify the wrong-gradient claim.

Initially, following your suggestion, I applied the same row_start correction as seen in the forward kernel to the pdist_backward_kernel_cuda_impl in https://github.com/pytorch/pytorch/pull/188006/commits/0cb6f555f947262a48920ad6ba183fc4ec723c9e. 

I then added a backward test in https://github.com/pytorch/pytorch/pull/188006/commits/c7257c803894ee8e6f9b06c9045ce4c6459609c1. This tests the boundaries of n /in {4096, 4097, 6000, 8000 and 10000} with atol / rtol set at 1e-6. 

I also manually experimented with a combination of $n \in \[50000, 70000]\$, $m \in \[1, 3]\$ and $p \in \[1.0, 2.0, inf]\$. The max diff here between the CPU and GPU is ~1e-11, which leads me to think it is FP64 rounding noise -- not numerical error.

However, with the above tests, I cannot reproduce a failure with or without the row_start edits in the `pdist_backward_kernel_cuda_impl` -- both cases pass as follows:

```
test/test_torch.py::TestTorchDeviceTypeCPU::test_pdist_norm_backward_index_cpu SKIPPED [0.0016s] (Only runs on cuda)                                 [ 50%]
test/test_torch.py::TestTorchDeviceTypeCUDA::test_pdist_norm_backward_index_cuda PASSED [0.7832s]                                                    [100%]

====================================================== 1 passed, 1 skipped, 1821 deselected in 2.49s =======================================================
```

This evidence suggests the "off by 1" issue does not occur in the backward pass's index arithmetic. Hence I cannot justify making this change and have reverted the modifications to pdist_backward_kernel_cuda_impl and the backward test.

If you have ideas for a repro, I can test and report back.

3. I accept. I have chosen to size the cap from `getCurrentDeviceProperties()->multiProcessorCount` rather than the magic 4096. Note this now applies to both backends. If you'd prefer zero change to the original NVIDIA launch, I can gate the cap under `#if defined(USE_ROCM)` and leave `dim3 grid(result.numel())` on CUDA — happy to do either.

4. I accept. `__syncthreads()` is now dropped and repeated experiments indicate correctness vs the CPU is unchanged.

Following the changes detailed in this comment, I have also modified the PR body accordingly. Please review and let me know.

**Comment by jeffdaily** [comment]
@geozhai I pushed a maintainer commit (39bba3f) on top of your branch rather than opening a competing PR, so your commits and this PR's history stay intact. It reduces the change to the grid fix, extends that fix to cdist, and rewrites the regression test. Details and the data behind each call below.

> **Root cause.** The grid limit is the whole story. `pdist_forward_kernel_impl` launches one 256-thread block per output, and HSA expresses a grid dimension in work-items rather than workgroups (`rocminfo` reports a per-dimension `Grid Max Size` of 4294967295), so `gridDim.x * blockDim.x` must fit in uint32. That caps the launch at 2^24 outputs. Bisected on gfx90a and gfx950 under HIP 7.14: n=5793 (16776528 blocks) works, n=5794 (16782321) fails. CUDA's `maxGridSize[0]` is in blocks, so the same launch is legal there.
>
> **Failure mode differs by HIP runtime, not by architecture.** You saw silent wrong results with unwritten outputs on MI355X under HIP 7.2. On the same MI355X under HIP 7.14 the oversized dispatch is rejected outright with `hipErrorInvalidConfiguration`. Same bug, same fix; the runtime just got stricter about validating it. Worth recording since the two symptoms look unrelated.
>
> **Why the index correction came out.** A perturbation of the fp64 sqrt propagates 1:1 into `n2 - sqrt(...)`, so it can only change `i` if that value falls within the perturbation of an integer. The smallest distance to an integer over every `k` is 1.2e-4 at n=4097, against a sqrt ulp of 9.1e-13: a margin of 134 million ulps. The tightest case up to n=50000 still has 1.4 million ulps of headroom. A ~1-ulp sqrt difference cannot move `i` by a row; n would have to reach ~1e10 before it could. That is a property of the arithmetic, so it holds under ROCm 7.2 as well, which is why I did not treat the n=4097 row as needing a separate fix. It also did not reproduce here on either gfx90a or gfx950, where fp32 pdist matches CPU exactly at n=4095/4096/4097/4098.
>
> The correction was also not free. Running the `row_start` fixup loop per output cost 1.6x to 1.8x on mid-size inputs (MI355X, 10-iteration averages):
>
> | case | main | PR as submitted | grid fix only |
> |---|---|---|---|
> | pdist n=4096 m=128 | 4.220 ms | 6.670 ms | 4.283 ms |
> | pdist n=4096 m=8 | 3.683 ms | 6.468 ms | 3.928 ms |
> | pdist n=2048 m=128 | 1.063 ms | 1.676 ms | 1.082 ms |
>
> Since the grid cap applies to NVIDIA too, that regression would have shipped everywhere.
>
> **Why `DeviceSqrt.cuh` was reverted.** `Normalization.cuh` (lines 250, 271, 412, 493) calls `device_sqrt` with no explicit template argument. On ROCm that resolves to the `float` specialization today, so making `device_sqrt` always return `double` silently moves ROCm batch-norm invstd from an fp32 to an fp64 divide. Unrelated to pdist and worth its own PR if it is wanted.
>
> **cdist had the identical bug.** `cdist_kernel_impl` uses the same `dim3 grid(result.numel())` with the same 256-thread block, and still failed at exactly 16777216 outputs after this PR's fix. It now shares a `dist_grid()` helper and grid-strides the same way. Reachable with any p that avoids the matmul path; a 4096x4096 cdist at p=3 is enough.
>
> **Test changes.** The comment on `test_pdist_norm_index_and_grid` recorded the fp32 `device_sqrt` and "~40% unwritten" explanation, which would have preserved a root cause that does not hold on either arch tested. Rewritten to state the work-item limit, moved to `@parametrize` for stable test ids, repinned from n=4097 to n=5793/5794 so it brackets the real boundary, and extended with cdist cases.
>
> **Validation** on gfx950 (MI355X, HIP 7.14): `test_torch.py -k "pdist or cdist"` 27 passed / 9 skipped; `test_ops.py -k "cdist or pdist"` 81 passed / 15 skipped / 5 xfailed; `test_autograd.py test_nn.py -k "cdist or pdist or pairwise"` 31 passed / 2 xfailed. `test_pdist_norm_large_cuda` passes in 4.7s. Forward also checked against CPU for p in {0, 0.5, 1, 2, 3, inf} above the old boundary, for batched cdist, and for degenerate shapes.
>
> One CI note: `test_pdist_norm_large` cannot be exercised via `PYTORCH_TEST_WITH_ROCM=1`. `TEST_WITH_ROCM` is derived from `torch.version.hip` regardless of that variable, so `@skipIfRocm` always skipped it; I ran the body directly to confirm the pre-fix failure.

The analysis above was drafted with an AI assistant (Claude Code) and reviewed by me. The measurements were run on an MI355X (gfx950) and MI250X (gfx90a) under ROCm 7.14.

**Review by jeffdaily — APPROVED** [review]
